### PR TITLE
DEPR/CLN: Remove pd.rolling_*, pd.expanding* and pd.ewm*

### PR DESCRIPTION
**Pull request by topper-123**
- [x] xref #18601, #18668 and #11603
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This PR removes ``pd.rolling_*``,`` pd.expanding_*`` ``pd.ewm*`` and the related ``pd.stats`` subpackage from the code base.

I have some doubts about some stuff:
* there are modules ``doc.plots.stats.*`` that I'm not sure of if they should be changed or removed. Currently they're not touched.
* In ``asv_bench.benchmarks.gil.py`` there are calls to ``pd.rolling_*`` etc. I've changed this to be method-based.
* ``pandas.tests.test_windows.py`` I felt was a bit complex to clean up.

I would appreciate inputs.

### CODE REVIEW

File: doc/source/whatsnew/v0.22.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: don’t remove any issue numbers

File: doc/source/whatsnew/v0.22.0.txt
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: don’t change other things

File: pandas/tests/test_window.py
Comment (jreback) [inline pandas/tests/test_window.py]: don’t remove tests just because we are catching the warning

you may need to rewrite them to use the new syntax
unless they duplicate current tests

**Review by jreback — CHANGES_REQUESTED** [review]
you are removing a lot of tests that should remain

File: doc/source/whatsnew/v0.22.0.txt
Comment (topper-123) [reply doc/source/whatsnew/v0.22.0.txt]: Hmm, this and some other stuff that’s unrelated to the pd.rolling stuff seems to be a rebase issue. I’ll fix it tomorrow.

If you want, you can limit the review to only the latest commit, the other commits are because of the rebase issue.

**Comment by topper-123** [comment]
The rebase issue has been solved now. I'll look at your comments now,. If you've got additional comments, I'd be thakful, and I'll look at them

**Comment by topper-123** [comment]
Hi @jreback. An issue is that dask apparently still uses ``pd.rolling_*``. This means that ``pandas/tests/test_downstream.py`` currently fails. Should this PR be put on hold, or would notifying the dask guys about this removal be enough? This stuff has been deprecated since 0.18, so for quite a long time.

This PR doesin't touch too many files, so IMO it would be ok to wait a bit, but given simplification to the pandas API, this PR should be included when 0.22 releases.

I've updated the PR, so grateful for a review, so it'll be ready for an eventual pull.

**Comment by jreback** [comment]
@TomAugspurger @mrocklin @jcrist

**Comment by TomAugspurger** [comment]
The dask versions are deprecated, just like pandas. I believe the best solution would be to

1. Use this PR to update all the test to the new syntax
2. Update dask to remove the `dd.rolling`, etc. methods (@topper-123 could I convince you to work on that? :) It'd be nice to have you as a contributor if you haven't already.)
3. Remove them from pandas once dask has a release

Depending on the timing of pandas 0.22, dask may not have a release out.

**Comment by topper-123** [comment]
@TomAugspurger , the ``test_dask`` function doesn't explicitly call ``pd.rolling_*``, so changes must be made in the dask code base, correct?

I've never used dask, and have a feeling I'd have to learn some intricancies of that lib to make such a PR, and I could easily go down a rabbit's hole that I never intended to enter... Could someone familiar with dask not take it upon themselves to do those changes? I can sit on this PR until dask is updated, and rebase when this is ready to merge.

**Comment by TomAugspurger** [comment]
@topper-123 dask was importing them at the top level if `dask.dataframe`, which caused the failure. Next time dask wants to mirror a pandas deprecation, we'll want to put it in a `try` / `except` block.

https://github.com/dask/dask/pull/2995

**Comment by topper-123** [comment]
@TomAugspurger , thanks for picking up on this, I will wait with my PR a while, until dash is updated. 

My PR lowers the number of top-level functions of pandas by some 30, so it would IMO be very beneficial to pandas' users to have this merged by 0.22, as it would make pandas seem much more focused. Could dask have a release before pandas 0.22, eventually just a minor release?

**Comment by topper-123** [comment]
dask 0.16.1 has now been released, and all tests now pass locally. The travis failure has to do with S3 and seems to be something about IO. That's the only failure in travis.

So I think everything is ok with my PR. Is there anything I can do to make this pass that travis error, or is this unrelated to my PR_

**Comment by topper-123** [comment]
All green.

**Comment by jorisvandenbossche** [comment]
Can you add a whatsnew notice?

File: pandas/tests/test_window.py
Comment (jreback) [inline pandas/tests/test_window.py]: you can parametrize this test on qs

File: pandas/tests/test_window.py
Comment (jreback) [inline pandas/tests/test_window.py]: there are 2 tests here (diff ddof)

File: pandas/tests/test_window.py
Comment (jreback) [inline pandas/tests/test_window.py]: where is the test for this one?

**Review by jreback — CHANGES_REQUESTED** [review]
small comments. looks pretty good.

File: pandas/tests/test_window.py
Comment (topper-123) [reply pandas/tests/test_window.py]: Not sure I understand, I'm replicating what was there before.

File: pandas/tests/test_window.py
Comment (topper-123) [reply pandas/tests/test_window.py]: Good catch, changed.

File: pandas/tests/test_window.py
Comment (jreback) [reply pandas/tests/test_window.py]: ahh I c, ok then.

File: pandas/tests/test_window.py
Comment (topper-123) [reply pandas/tests/test_window.py]: There's a test ``test_ewm_alpha`` that is the same as this.

Comment (jreback) [reply pandas/tests/test_window.py]: ok great

**Review by jreback — APPROVED** [review]
lgtm. ping on green.

**Comment by topper-123** [comment]
All green

File: pandas/tests/test_window.py
Comment (jorisvandenbossche) [inline pandas/tests/test_window.py]: Previously this function consisted of `_check_ew_ndarray` and `_check_ew_structures` subfunctions to test both interfaces. You removed `_check_ew_ndarray` and only kept the content of `_check_ew_structures` as this is testing the new interface. 
But, as far as I can see, the *actual* testing was done in `_check_ew_ndarray`  (`_check_ew_structures` is only testing the output class type). Shouldn't this testing be kept? (changed to the new interface)

(might good be I am missing something, or that it is covered in other tests)

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
Question about one of the tests

File: pandas/tests/test_window.py
Comment (topper-123) [reply pandas/tests/test_window.py]: I think you are right: So rather than test this function call, test the ``ewm`` method call on a series. I'll look into it tomorrow.

**Comment by topper-123** [comment]
@jorisvandenbossche , I've now replicated the code from ``_check_ew_ndarray`` into ``check_ew``. Could you review it?

**Comment by jorisvandenbossche** [comment]
@topper-123 that specific one looks good now I think. 
Did you double check the other tests to see there were not similar problems?

**Comment by topper-123** [comment]
I've added ``test_expanding_func`` (parametrized, was called ``test_parametrized_max`` previously) and ``test_expanding_apply``. Everything is now included AFAICT.

**Comment by topper-123** [comment]
The tests have been running for 9 hours now and the Travis-CI tests have not yet finished. Same thing happened yesterday where after 20 hours one travis test-suite failed without console output (my guess is it timed out).  Is anything wrong with the travis suite? 

Everything passes locally and the errors that travis reported two days ago have been fixed (but the fixes not confirmed by travis, because I can't get travis to go through them...).

What should I do? Could a core dev run the travis tests locally and confirm they're ok, and merge them?

**Comment by TomAugspurger** [comment]
TravisCI had an outage yesterday, so all the builds were canceled and rescheduled. Looks like yours is 4th on deck at the moment: https://travis-ci.org/pandas-dev/pandas/pull_requests

**Comment by topper-123** [comment]
Ok, thanks. I'll wait a bit more then.

**Comment by topper-123** [comment]
All green now🤓

**Comment by jreback** [comment]
when @jorisvandenbossche approves can merge.

**Comment by topper-123** [comment]
@jorisvandenbossche, I'd really appreciate if you could do this.

File: pandas/tests/test_window.py
Comment (jorisvandenbossche) [inline pandas/tests/test_window.py]: Do we need to move parts of `_check_ndarray` like this part on checking `min_periods` to `_check_moments_func` ? 

From a superficial look, it seems that `_check_ndarray` did more than `_check_structures` (eg in `_check_structures` I don't see any assertion of NaNs when `min_periods` is specified like here)

I also don't know if `test_stable` and `test_window` make sense for the new interface

**Review by jorisvandenbossche — COMMENTED** [review]
sorry @topper-123, one more question 
(it's a very complex file / diff (not your fault!) to review)

**Comment by topper-123** [comment]
Changes suggested by @jorisvandenbossche implemented and all green.

File: pandas/tests/test_window.py
Comment (jreback) [inline pandas/tests/test_window.py]: I am not sure of the point of adding ``_check_series``, it *seems* that you are already checking the series & frame results (line 1344), yes?

Comment (jreback) [reply pandas/tests/test_window.py]: if there IS a difference, IOW you are checking something different, then certainly don't make a new function, just in-line it here.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by topper-123** [comment]
Hi @jreback. I uploaded a new PR 12 days ago.

There were a few issues that were checked in ``_check_ndarray`` but were not in ``_check_moments`` previously:
* tests for whether Nans were excluded correctly
* tests for whether ``center`` worked correctly
* tests for what happens when the window is larger than the series

All of the above are inlined in ``_check_moments`` now (see 37abd2582b5a1043fe460f8793c4480116df31f3).

I've rebased a moments ago, without making code changes otherwise, so everything should turn up green & reased when tests are finished.

File: pandas/tests/test_window.py
Comment (jreback) [inline pandas/tests/test_window.py]: hmm, shouldn't this be nansum?

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine. just that 1 question.

File: pandas/tests/test_window.py
Comment (topper-123) [reply pandas/tests/test_window.py]: Yes. fixed.

**Comment by topper-123** [comment]
all green.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@topper-123 Thanks a lot for this!

**Pull request merged**
